src/clib/pio_darray.c

```diff
@@ -1357,32 +1359,37 @@ int flush_output_buffer(file_desc_t *file, bool force, PIO_Offset addsize)
     maxusage = usage;
   }
   if(force || usage>=PIO_BUFFER_SIZE_LIMIT){
-    int request[2*PIO_MAX_VARS];
-    int status[2*PIO_MAX_VARS];
-    int rcnt=0;
+    int rcnt;
     bool prev_dist=false;
     int prev_record=-1;
     int prev_type=0;
-
+    int  maxreq;
+    int reqcnt;
+    maxreq = 0;
+    reqcnt=0;
+    rcnt=0;
     for(int i=0; i<PIO_MAX_VARS; i++){
       vdesc = file->varlist+i;
+      reqcnt+=vdesc->nreqs;
+      if(vdesc->nreqs>0) maxreq = i;
+    }
+    int request[reqcnt];
+    int status[reqcnt];
 
-
+    for(int i=0; i<=maxreq; i++){
+      vdesc = file->varlist+i;
 #ifdef MPIO_ONESIDED
```

Comment:
Early this year I worked on IO performance with ROMIO (MPI-IO) and PnetCDF developers.
The MPIO_ONESIDED optimization went into the ROMIO library - that optimization should eventually make it into the mpich distribution but as far as I know that hasn't happened yet.  
It improves performance but introduces restrictions on what you can put into the write buffer - 
the buffer (aggregated across IO tasks) must represent a contiguous block on the file.  This ifdef is about trying to assure that that is the case.